src/lib/Transform.ts

```diff
@@ -185,37 +195,43 @@ export function batch<T>(
 
 	readable.forEach(
 		async (v: T): Promise<void> => {
-			throwIfThrowable(consumeEarlyEndError());
-
+			startFlushTimeout();
 			queue.push(v);
+			let flushFailureError: Error | undefined;
+			let earlyFlushError: Error | undefined;
 
 			if (queue.length >= maxBatchSize) {
-				// backpressure
-				await flush();
-			} else if (queue.length >= minBatchSize && !pendingWrite) {
+				try {
+					// backpressure
+					earlyFlushError = await settleEarlyFlush();
+					await flush();
+				} catch (e) {
+					flushFailureError = e;
+				}
+			} else if (!pendingWrite) {
 				// no backpressure yet (until new queue fills to maxBatchSize)
-				earlyFlush();
+				swallowErrors(earlyFlush());
 			}
 
-			if (queue.length && flushTimeout !== undefined && !pendingWrite) {
-				if (timer !== undefined) {
-					clearTimeout(timer);
-				}
-				timer = setTimeout(earlyFlush, flushTimeout);
-			}
+			throwIfThrowable(earlyFlushError);
+			throwIfThrowable(consumeEarlyFlushError()); // Catch the error early if possible
+			throwIfThrowable(flushFailureError);
 		},
 		async (error?: Error) => {
-			let toThrow = consumeEarlyEndError() || error;
+			clearFlushTimeout();
+			let flushError: Error | undefined;
+
+			const earlyFlushError = await settleEarlyFlush();
 
 			try {
 				await flush();
 			} catch (e) {
-				toThrow = e;
+				flushError = e;
 			}
 
-			await writable.end(toThrow, readable.result());
-
-			throwIfThrowable(toThrow);
+			await writable.end(error, readable.result());
```

Comment:
The reason for the failing tests is that `readable.result()` resolves to an error. You can check it by e.g.:
```
			readable.result().then(
				() => console.log("RESULT OK"),
				(err) => console.log("RESULT ERR", err)
			);
```

(Printf debugging, yay!)

My time for today is up unfortunately :(
You'll have to dig into the source side to see why they apparently set their result to fail.

Reply:
The batcher itself throws any remaining errors on end, and `_pump()` listens for errors from `ender()` as part of `result()` as shown here:

```
this._readBusy = track(
    Promise.resolve(eof).then((eofValue) => ender(enderArg))
);
```

```
    ...
    p = this._readBusy.promise;
}
this._resultDeferred.resolve(p);
```

Under a more normal one-to-one stream this might be fine because if a stream is ending, its downstream is also ending and anything that's waiting for a downstream result may also want to know about an error in an upstream ender.

In this case, though, the flush errors need to go upstream, to the caller of `end()`, and not downstream via `result()`. If they go downstream then the source cannot decide what to do with them; plus it's inconsistent with the other case, which is a `flush()` `await`ed inside the reader.

Reply:
Tried
```
await writable.end(error, readable.result().catch(noop));
```
It worked, but it feels monstrous.

Reply:
(Side note: VSCode's `async`/`await` debugging seems really brittle, I've had to rely extremely heavily on logging.)

Reply:
> It worked, but it feels monstrous.

I agree. This should not be the way to go. You make good points about the direction of error flow.
The existing behavior was put it for a reason, and changing it breaks several other tests.
I've tinkered a bit with it this evening, but didn't come with a satisfactory solution yet.

There are several options:
1. Use your workaround
2. Remove throwing the buffered errors
3. Do not let an ender error be 'forwarded' downstream, except if upstream didn't really handle it.

Now, option 3 would be the best, *however* it may not be possible to do this in practice given behavior we need in several other places. It will definitely require a few changes in the core, and some of the existing transforms, and would then be a breaking change.

I'll need to think this piece over for a bit. I'll keep you posted.

> (Side note: VSCode's async/await debugging seems really brittle, I've had to rely extremely heavily on logging.)

Yes, I've noticed that too. It seems to work fine as long as you only use 'proper' `await`, but as soon as you throw in a 'normal' `promise.then()` (which is what ts-stream does everywhere, as async/await didn't exist at the time), it can't follow the traces anymore. I feel your pain...

Reply:
Not to interrupt your thinking, but there's another solution I'd like to suggest as well:
4. Provide a new optional parameter, `handleError`, that is called whenever the batcher's `write()` call fails.

This gives the upstream the opportunity to work directly with errors that occur outside of its `write()` promise, letting them choose how to handle any `async` quirks. If their handler throws, or if the handler is not provided, then we fall back to the current behaviour.

And/or:
5. In the case of `batch()`, *always* throw errors downstream, even if they occurred due to a rejected promise of `forEach()`.

I suggest this not because I like it, but for these reasons:

- It requires no changes to the core
- Even without the fancy options of `minBatchSize` and `flushTimeout`, `batch()` would have this problem anyway because the stream needs to execute a final `flush()` on `end()`
- It hides the implementation detail of whether a `flush()` operation occurs inside `forEach()` or separately from it
- Even if bad, it is consistent, which makes it easier to document and makes bugs resulting from it easier to trace.

Implementing 5) might be more acceptable if we also implement 4), because then a user can avoid the pain of the weird behaviour by providing an error handler.

Reply:
That optional `handleError` is an interesting thought too, and I think it would indeed help especially with the minBatchSize flushes. We may want to implement that in any case.
Always throwing downstream would contradict with the thinking behind the library, so I'd rather prevent that.

Btw, I'm not opposed to making a change to the core, especially if the existing behavior can actually considered to be more of a bug. I do want to have/keep sensible behavior around the other transforms such as map and filter, so if those would become less 'intuitive' due to this change, I won't do it.

Reply:
If it's possible, I like 3) the best. Thinking about how to do it, maybe something like:
```
const enderPromise = track( Promise.resolve(eof).then((eofValue) => ender(enderArg)) );
const enderPromiseCatch = enderPromise.promise.catch.bind(enderPromise);
this.enderPromiseCaught = false;
enderPromise.promise.catch = (e) => { this.enderPromiseCaught = true; return enderPromiseCatch(e); }

this._readBusy = enderPromise;
```

And then elsewhere:
```
this._resultDeferred.resolve(p.catch(e => { if (!this.enderPromiseCaught) { throw e; } });
```

That's assuming that `async/await` `catch` actually calls `catch()` on the promise it catches.

I don't know if this would work. The obvious problem is that the catcher of `enderPromise` from `end()` might just re-throw the error. Maybe that's fine. I can play around with this tomorrow and see how it behaves.

Reply:
Another, only slightly less hacky fix, this time in `batch()`: if an error will throw from the ender of `batch()`, we catch it and then swallow it in the promise we pass downstream:

```
writableStream.end(error, readable.result().catch(e => if (e !== toThrow) { throw e; }));
```

This will suppress the ender error from `result()`, but leave any other errors alone.

Reply:
> Btw, I'm not opposed to making a change to the core, especially if the existing behaviour can actually considered to be more of a bug.

On this note, I am interested in possibly extending the behaviour of `abort()` to allow a "clean" abort (like the behaviour we talked about in issue #47 ). If I did that, how would you prefer it done? A breaking change to the API, a new method and deprecation of `abort()`, etc. ...?

Reply:
I was working on it yesterday a little bit. Some of the more exceptional cases around errors in enders etc are actively being used and tested for, but apparently mostly indirectly through `map`, `filter` and `writeEach` tests.

So my first steps are to add some tests around the low-level primitives themselves (such as `end()`) so explicitly specify 
what they should be doing in these cases, and then check whether it's possible to (cleanly) adapt the other functions to the new situation.

I only have limited time available this weekend, so it may take a while.

> On this note, I am interested in possibly extending the behaviour of abort() to allow a "clean" abort (like the behaviour we talked about in issue #47 ). If I did that, how would you prefer it done? A breaking change to the API, a new method and deprecation of abort(), etc. ...?

Let's open a new issue to discuss that, because this one is complex enough as it is :)
In any case, I'd first want to discuss the design options before making any actual code changes.